File: android/app/src/main/java/com/mwiacek/przepisy/drogowe/ScrollerView.java
package com.mwiacek.przepisy.drogowe;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.os.CountDownTimer;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

//Shared between Straz and Przepisy
//1.0
public class ScrollerView extends View {
    boolean Touched = false, Showed = false;
    int y;
    Paint p = new Paint();
    Bitmap myBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.scroll1);
    CountDownTimer ct;
    ScrollerWebView webview;

    public ScrollerView(Context context) {
        super(context);
        p.setAlpha(200);
    }

    public ScrollerView(Context context, AttributeSet attrs) {
        super(context, attrs);
        p.setAlpha(200);
    }

    public ScrollerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        p.setAlpha(200);
    }

    public void SetTimer() {
        if (ct != null) ct.cancel();
        Showed = true;
        invalidate();
        ct = new CountDownTimer(2000, 2000) {
            public void onTick(long millisUntilFinished) {
            }

            public void onFinish() {
                if (Showed || !Touched) {
                    Showed = false;
                    invalidate();
                }
            }
        }.start();
    }

    @Override
    public void onDraw(Canvas canvas) {
        if (Showed) canvas.drawBitmap(myBitmap, 0, y, p);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (!Showed) return false;
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            if ((int) event.getY() >= y && (int) event.getY() <= y + myBitmap.getHeight()) {
                Touched = true;
                return true;
            }
        }
        if (event.getAction() == MotionEvent.ACTION_MOVE && Touched) {
            y = (int) event.getY();
            if (y + myBitmap.getHeight() > webview.getHeight()) {
                y = webview.getHeight() - myBitmap.getHeight();
            }
            if (y - myBitmap.getHeight() < 0) y = 0;
            webview.scrollTo(0,
                    (y * webview.GetMax() / (this.getHeight() - myBitmap.getHeight())));
            return true;
        }
        if (event.getAction() == MotionEvent.ACTION_UP && Touched) {
            Touched = false;
            SetTimer();
            return true;
        }
        return false;
    }
}

File: android_kotlin/app/src/main/java/com/mwiacek/przepisy/drogowe/ScrollerView.kt
package com.mwiacek.przepisy.drogowe

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint
import android.os.CountDownTimer
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

//Shared between Straz and Przepisy
//1.0
class ScrollerView : View {
    var Touched = false
    var Showed = false

    @JvmField
    var y = 0
    var p = Paint()

    @JvmField
    var myBitmap = BitmapFactory.decodeResource(resources, R.drawable.scroll1)
    var ct: CountDownTimer? = null
    var webview: ScrollerWebView? = null

    constructor(context: Context?) : super(context) {
        p.alpha = 200
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        p.alpha = 200
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    ) {
        p.alpha = 200
    }

    fun SetTimer() {
        if (ct != null) ct!!.cancel()
        Showed = true
        invalidate()
        ct = object : CountDownTimer(2000, 2000) {
            override fun onTick(millisUntilFinished: Long) {}
            override fun onFinish() {
                if (Showed || !Touched) {
                    Showed = false
                    invalidate()
                }
            }
        }.start()
    }

    public override fun onDraw(canvas: Canvas) {
        if (Showed) canvas.drawBitmap(myBitmap, 0f, y.toFloat(), p)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (!Showed) return false
        if (event.action == MotionEvent.ACTION_DOWN) {
            if (event.y.toInt() >= y && event.y.toInt() <= y + myBitmap.height) {
                Touched = true
                return true
            }
        }
        if (event.action == MotionEvent.ACTION_MOVE && Touched) {
            y = event.y.toInt()
            if (y + myBitmap.height > webview!!.height) {
                y = webview!!.height - myBitmap.height
            }
            if (y - myBitmap.height < 0) y = 0
            webview!!.scrollTo(
                0,
                y * webview!!.GetMax() / (this.height - myBitmap.height)
            )
            return true
        }
        if (event.action == MotionEvent.ACTION_UP && Touched) {
            Touched = false
            SetTimer()
            return true
        }
        return false
    }
}